### DISCUSSION TITLE
argument type that is constrained to a class and protocol

### DISCUSSION THREAD

**Discussion by shadycuz** [python/typing#1881] [Q&A] [answered]
Not sure if I explained it correctly in the title, so here is some sample code.
```python3
from typing import Protocol, TypeVar


class HasJob(Protocol):

    def job(self) -> str: ...


class Person:
    def __init__(self, name: str, age: int):
        self.name = name
        self.age = age


class Engineer(Person):
    def __init__(self, name: str, age: int):
        super().__init__(name, age)

    def job(self) -> str:
        return "Engineer"


# T = TypeVar("T", bound=HasJob)


# A function that accepts a Person with a job method
def go_to_work(person: Person):
    print(f"{person.name} is going to work as a {person.job()}")


bob = Engineer("Bob", 30)

go_to_work(bob)

```

I would like my `go_to_work` function to take any `Person` or subclass of `Person`. But the class that is passed in must also implement `HasJob`. 

I have not found a way to do it with `TypeVar` as is. I'm trying to avoid creating a new class like `Employee`.

**Comment by JelleZijlstra** [comment, marked as answer]
What you're looking for is called an intersection type. This is not currently part of the Python type system, but it's a popular idea: #213.

**Reply by shadycuz** [reply]
Thanks for linking to that. Since they are asking for people not to post there, then I guess I will post here =). 

Even though you cant specify a intersection type today. If you take a class and then check if it fits a protocol using `isinstance`, Pylance can put together the correct "intersection" type. 

Code:
```python3
if isinstance(compatiple_version, UpgradeableConfig):
  params = compatiple_version.upgrade(params)
 ```
Pylance says the type is:
<img width="725" alt="image" src="https://github.com/user-attachments/assets/ca0cbafe-e036-49f3-96ca-94bf083213f9">

It does seem like, at least in Pylance, the intersection is possible to have. It's just not possible to define one.

**Reply by JelleZijlstra** [reply]
Yes, type checkers may internally synthesize an intersection type in some cases. However, it is not possible to define such a type in an annotation.
